File: happy_numbers.rb
# http://rosettacode.org/wiki/Happy_numbers

def happy?(number)
  while (number != 1) && (number != 89)
    number = sum_square_digits(number)
  end
  number == 1
end

def sum_square_digits(number,accumulator=0)
  number < 10 ? accumulator + (number ** 2) : ((number % 10) ** 2) + sum_square_digits(number/10,accumulator)
end

def find_happy_numbers(number,array=[],counter=1)
  until array.count == number
    array << counter if happy?(counter) 
    counter += 1
  end
  array
end  

p find_happy_numbers(8)

File: happy_numbers.swift
// http://rosettacode.org/wiki/Happy_numbers

func happy(n:Int) -> Bool {
  var result = sumSquareDigits(n,0)
  while (result != 89) && (result != 1) {
    result = sumSquareDigits(result,0)
  }
  if result == 1 {
    return true
  }
  return false
}

func sumSquareDigits(n:Int,accumulator:Int) -> Int {
  var mutableAccumulator = accumulator
  if n < 10 {
    return mutableAccumulator + (n * n)
  } else {
    return ((n % 10) * (n % 10)) + sumSquareDigits(n/10,(mutableAccumulator))
  }
}

var happyNumbers:[Int] = []
var counter = 1
while happyNumbers.count < 8 {
  if happy(counter) {
    happyNumbers.append(counter)
  }
  counter++
}

println(happyNumbers)
